cgmes/cgmes-completion/src/main/java/com/powsybl/cgmes/completion/CreateMissingContainersPreProcessor.java

```diff
@@ -0,0 +1,216 @@
+            return;
+        }
+        Path fixesFile = fixesFolder.resolve(basename + ".zip");
+        // Check the file will be writable
+        try {
+            Files.deleteIfExists(fixesFile);
+            Files.createFile(fixesFile);
+        } catch (IOException e) {
+            LOG.error("Output file {} is not writable. Skipping post processor.", fixesFile);
+            return;
+        }
+        if (LOG.isInfoEnabled()) {
+            LOG.info("Execute {} post processor on CGMES model {}. Output to file {}", getName(), cgmes.modelId(), fixesFile);
+        }
+        buildFixesOnZipFile(cgmes, basename, fixesFile);
+    }
+
+    private static void buildFixesOnZipFile(CgmesModel cgmes, String basename, Path fixesFile) {
+        // Assume all containers missing are voltage levels and create proper objects for them (substation, regions, ...)
+        Set<String> missingVoltageLevels = findMissingVoltageLevels(cgmes);
+        LOG.info("Missing voltage levels: {}", missingVoltageLevels);
+        if (!missingVoltageLevels.isEmpty()) {
+            buildZipFileWithFixes(missingVoltageLevels, fixesFile, basename);
+            cgmes.read(new ZipFileDataSource(fixesFile), Reporter.NO_OP);
+        }
+        Set<String> missingVoltageLevelsAfterFix = findMissingVoltageLevels(cgmes);
+        if (!missingVoltageLevelsAfterFix.isEmpty()) {
+            throw new IllegalStateException("Missing voltage levels after fix: " + missingVoltageLevelsAfterFix);
+        }
+        // The only containers without voltage level must be of type line
+        LOG.info("After the fixes have been applied, the only node containers without voltage level must be of type Line.");
+        LOG.info("Containers without voltage level that are not Lines will be reported as errors.");
+        cgmes.connectivityNodeContainers().stream()
+                .filter(c -> c.getId(CgmesNames.VOLTAGE_LEVEL) == null)
+                .filter(c -> !c.getLocal("connectivityNodeContainerType").equals("Line"))
+                .forEach(c -> LOG.error(c.getId(CgmesNames.CONNECTIVITY_NODE_CONTAINER)));
+    }
+
+    private static Set<String> findMissingVoltageLevels(CgmesModel cgmes) {
+        // check missing CN containers
+        Set<String> defined = cgmes.connectivityNodeContainers().stream().map(c -> c.getId(CgmesNames.CONNECTIVITY_NODE_CONTAINER)).collect(Collectors.toSet());
+        Set<String> referred = cgmes.connectivityNodes().stream().map(c -> c.getId(CgmesNames.CONNECTIVITY_NODE_CONTAINER)).collect(Collectors.toSet());
+        return referred.stream().filter(c -> !defined.contains(c)).collect(Collectors.toSet());
+    }
+
+    private static void buildZipFileWithFixes(Set<String> missingVoltageLevels, Path fixesFile, String basename) {
+        Network network = NetworkFactory.findDefault().createNetwork("empty", "CGMES");
+        // TODO(Luma) should we ensure that context CIM version uses the same version of the input files ?
```

Comment:
Can you elaborate this `TODO`?

Reply:
We have read CGMES files in a given version.
We should export CGMES files in the same version.
When the network is created applying conversion from CGMES files, the version of input data is stored in the extension `CimCharacteristics`, and when an export context is created from a `Network`, the version to use in the export is the same one that was read, unless explicitly changed.
Here we are creating an export context with an empty `Network` that will not carry information about specific CIM version to use, so the default value will be used (CIM16).

I will set the CIM version in the context explicitly, based on the version of the input files.